File: Appstax/Appstax/Appstax.m

#import "AppstaxInternals.h"
#import "AXJsonApiClient.h"
#import "AXDataStore.h"

@implementation Appstax

- (void)setupServicesWithAppKey:(NSString *)appKey {
    [self setupServicesWithAppKey:appKey baseUrl:@"https://appstax.com/api/latest/"];
}

- (void)setupServicesWithAppKey:(NSString *)appKey baseUrl:(NSString *)baseUrl {
    _appKey = appKey;
    _apiClient = [[AXJsonApiClient alloc] initWithAppKey:appKey baseUrl:baseUrl];
    [self setupServicesWithApiClient:_apiClient];
}

- (void)setupServicesWithApiClient:(AXJsonApiClient *)apiClient {
    _apiClient = apiClient;
    _dataStore = [[AXDataStore alloc] initWithApiClient:_apiClient];
    _userService = [[AXUserService alloc] initWithApiClient:_apiClient];
    _permissionsService = [[AXPermissionsService alloc] initWithApiClient:_apiClient];
    _fileService = [[AXFileService alloc] initWithApiClient:_apiClient];
}

#pragma mark - Class methods

+ (void)setAppKey:(NSString *)appKey {
    [[self defaultContext] setupServicesWithAppKey:appKey];
}

+ (void)setAppKey:(NSString *)appKey baseUrl:(NSString *)baseUrl {
    [[self defaultContext] setupServicesWithAppKey:appKey baseUrl:baseUrl];
}

+ (NSString *)appKey {
    return [[self defaultContext] appKey];
}

+ (void)setApiClient:(AXJsonApiClient *)apiClient {
    [[self defaultContext] setupServicesWithApiClient:apiClient];
}

+ (instancetype)defaultContext {
    static Appstax *defaultContext = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        defaultContext = [[self alloc] init];
    });
    return defaultContext;
}

+ (NSBundle *)frameworkBundle {
    static NSBundle* frameworkBundle = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        frameworkBundle = [NSBundle bundleForClass:[Appstax class]];
    });
    return frameworkBundle;
}

+ (void)includeUnusedClasses {
    [AXImageView class];
}

@end

File: Appstax/Appstax/Appstax.swift

import Foundation

@objc public class Appstax: NSObject {
    
    private(set) public var apiClient: AXApiClient!
    private(set) public var appKey: String = ""
    private(set) public var fileService: AXFileService!
    private(set) public var objectService: AXObjectService!
    private(set) public var userService: AXUserService!
    private(set) public var permissionsService: AXPermissionsService!
    private(set) var realtimeService: AXRealtimeService!
    
    public static func setAppKey(appKey: String) {
        Appstax.defaultContext.setupServicesWithAppKey(appKey)
    }
    
    public static func setAppKey(appKey: String, baseUrl: String) {
        Appstax.defaultContext.setupServicesWithAppKey(appKey, baseUrl: baseUrl)
    }
    
    public static func setLogLevel(levelName: String) {
        if let level = AXLog.levelByName(levelName) {
            AXLog.minLevel = level
        }
    }
    
    internal func setupServicesWithAppKey(appKey: String, baseUrl: String = "https://appstax.com/api/latest/") {
        self.appKey = appKey
        self.apiClient = AXApiClient(appKey: appKey, baseUrl: baseUrl)
        setupServicesWithApiClient(apiClient!)
    }
    
    internal func setupServicesWithApiClient(apiClient: AXApiClient) {
        self.apiClient = apiClient
        self.objectService = AXObjectService(apiClient: apiClient)
        self.userService = AXUserService(apiClient: apiClient)
        self.permissionsService = AXPermissionsService(apiClient: apiClient)
        self.fileService = AXFileService(apiClient: apiClient)
        self.realtimeService = AXRealtimeService(apiClient: apiClient)
        AXLog.info("Initialized Appstax with app key \(appKey) and base url \(apiClient.baseUrl)")
    }
    
    public static func frameworkBundle() -> NSBundle! {
        struct Static {
            static var onceToken: dispatch_once_t = 0
            static var frameworkBundle: NSBundle? = nil
        }
        dispatch_once(&Static.onceToken) {
            Static.frameworkBundle = NSBundle(forClass: Appstax.self)
        }
        return Static.frameworkBundle
    }
    
    public static var defaultContext: Appstax {
        struct Static {
            static var onceToken: dispatch_once_t = 0
            static var instance: Appstax? = nil
        }
        dispatch_once(&Static.onceToken) {
            Static.instance = Appstax()
        }
        return Static.instance!
    }

}
